File: Protocol/_theoreme_sim/src/research/stabilityScorer.js
/* Purpose: Convert long-run simulation snapshots into comparable stability score and verdict. */

/**
 * @param {number[]} values
 * @returns {number}
 */
function average(values) {
  if (!values || values.length === 0) {
    return 0;
  }
  return values.reduce((sum, value) => sum + value, 0) / values.length;
}

/**
 * @param {number[]} values
 * @returns {{early:number,late:number,delta:number}}
 */
function trendDelta(values) {
  if (!values || values.length === 0) {
    return { early: 0, late: 0, delta: 0 };
  }

  const chunk = Math.max(1, Math.floor(values.length * 0.25));
  const early = average(values.slice(0, chunk));
  const late = average(values.slice(-chunk));
  return {
    early,
    late,
    delta: early - late,
  };
}

/**
 * @param {any[]} snapshots
 * @returns {{tailRatio:number,maxTail:number,minTail:number}}
 */
function tailEligibleHealth(snapshots) {
  if (!snapshots || snapshots.length === 0) {
    return { tailRatio: 0, maxTail: 0, minTail: 0 };
  }

  const chunk = Math.max(3, Math.floor(snapshots.length * 0.2));
  const tail = snapshots
    .slice(-chunk)
    .map((item) => Number(item.eligibleCount || 0));
  const maxTail = Math.max(...tail, 0);
  const minTail = Math.min(...tail);

  return {
    tailRatio: maxTail > 0 ? minTail / maxTail : 0,
    maxTail,
    minTail,
  };
}

/**
 * @param {{snapshots:any[],finalSnapshot:any}} run
 * @returns {{score:number,verdict:string,metrics:Record<string,number|string|boolean>,rationale:string[]}}
 */
export function scoreStability(run) {
  const snapshots = run.snapshots || [];
  const totalRounds = snapshots.length;

  if (totalRounds === 0) {
    return {
      score: 0,
      verdict: "UNSTABLE",
      metrics: {
        totalRounds: 0,
      },
      rationale: ["No rounds executed."],
    };
  }

  const theoremPassRate =
    snapshots.filter((item) => item.theoremPass === true).length / totalRounds;

  const assumptionsPassRate =
    snapshots.filter((item) => item.assumptionsPass === true).length /
    totalRounds;

  const coverage = snapshots.map((item) => Number(item.downCoverage || 0));
  const coverageAvg = average(coverage);

  const duplicates = snapshots.map((item) => Number(item.downDuplicates || 0));
  const duplicateTrend = trendDelta(duplicates);

  const parentChanges = snapshots.map((item) =>
    Number(item.parentChanges || 0),
  );
  const parentChangeAvg = average(parentChanges);

  const flapping = snapshots.map((item) => Number(item.flappingNodes || 0));
  const flappingAvg = average(flapping);

  const eligibleHealth = tailEligibleHealth(snapshots);

  const theoremScore = theoremPassRate * 30;
  const assumptionsScore = assumptionsPassRate * 18;
  const coverageScore = Math.min(1, coverageAvg / 0.95) * 18;
  const duplicateScore =
    duplicateTrend.delta > 0
      ? Math.min(1, duplicateTrend.delta / Math.max(1, duplicateTrend.early)) *
        16
      : 0;

  const parentScore =
    Math.max(0, 10 - Math.min(10, parentChangeAvg)) +
    Math.max(0, 8 - Math.min(8, flappingAvg * 1.5));

  let score =
    theoremScore +
    assumptionsScore +
    coverageScore +
    duplicateScore +
    parentScore;

  const rationale = [];

  if (eligibleHealth.tailRatio < 0.6) {
    score *= 0.7;
    rationale.push("Eligible-set collapse detected in tail rounds.");
  }

  if (
    snapshots.slice(-8).every((item) => Number(item.eligibleCount || 0) <= 1)
  ) {
    score *= 0.45;
    rationale.push(
      "Vacuous tail: only gateway remains eligible in final rounds.",
    );
  }

  if (flappingAvg > 2.2) {
    score *= 0.82;
    rationale.push("High average flapping detected.");
  }

  score = Math.max(0, Math.min(100, score));

  const verdict =
    score >= 80 ? "STABLE" : score >= 60 ? "OSCILLATING" : "UNSTABLE";

  return {
    score,
    verdict,
    metrics: {
      totalRounds,
      theoremPassRate,
      assumptionsPassRate,
      coverageAvg,
      duplicateEarly: duplicateTrend.early,
      duplicateLate: duplicateTrend.late,
      duplicateDrop: duplicateTrend.delta,
      parentChangeAvg,
      flappingAvg,
      eligibleTailRatio: eligibleHealth.tailRatio,
      eligibleTailMin: eligibleHealth.minTail,
      eligibleTailMax: eligibleHealth.maxTail,
    },
    rationale,
  };
}

File: Protocol/_theoreme_sim_py/src/research/stability_scorer.py
"""Purpose: Convert simulation snapshots into comparable stability score and verdict."""

from __future__ import annotations


def _average(values: list[float]) -> float:
    if not values:
        return 0.0
    return sum(values) / len(values)


def _trend_delta(values: list[float]) -> dict:
    if not values:
        return {"early": 0.0, "late": 0.0, "delta": 0.0}

    chunk = max(1, int(len(values) * 0.25))
    early = _average(values[:chunk])
    late = _average(values[-chunk:])
    return {"early": early, "late": late, "delta": early - late}


def _tail_eligible_health(snapshots: list[dict]) -> dict:
    if not snapshots:
        return {"tailRatio": 0.0, "maxTail": 0.0, "minTail": 0.0}

    chunk = max(3, int(len(snapshots) * 0.2))
    tail = [float(item.get("eligibleCount", 0)) for item in snapshots[-chunk:]]
    max_tail = max([0.0, *tail])
    min_tail = min(tail) if tail else 0.0

    return {
        "tailRatio": (min_tail / max_tail) if max_tail > 0 else 0.0,
        "maxTail": max_tail,
        "minTail": min_tail,
    }


def score_stability(run: dict) -> dict:
    snapshots = run.get("snapshots", [])
    total_rounds = len(snapshots)

    if total_rounds == 0:
        return {
            "score": 0,
            "verdict": "UNSTABLE",
            "metrics": {"totalRounds": 0},
            "rationale": ["No rounds executed."],
        }

    theorem_pass_rate = (
        len([item for item in snapshots if item.get("theoremPass") is True])
        / total_rounds
    )
    assumptions_pass_rate = (
        len([item for item in snapshots if item.get("assumptionsPass") is True])
        / total_rounds
    )

    coverage = [float(item.get("downCoverage", 0)) for item in snapshots]
    coverage_avg = _average(coverage)

    duplicates = [float(item.get("downDuplicates", 0)) for item in snapshots]
    duplicate_trend = _trend_delta(duplicates)

    parent_changes = [float(item.get("parentChanges", 0)) for item in snapshots]
    parent_change_avg = _average(parent_changes)

    flapping = [float(item.get("flappingNodes", 0)) for item in snapshots]
    flapping_avg = _average(flapping)

    eligible_health = _tail_eligible_health(snapshots)

    theorem_score = theorem_pass_rate * 30
    assumptions_score = assumptions_pass_rate * 18
    coverage_score = min(1.0, coverage_avg / 0.95) * 18
    duplicate_score = (
        min(1.0, duplicate_trend["delta"] / max(1.0, duplicate_trend["early"])) * 16
        if duplicate_trend["delta"] > 0
        else 0
    )

    parent_score = (
        max(0.0, 10 - min(10, parent_change_avg))
        + max(0.0, 8 - min(8, flapping_avg * 1.5))
    )

    score = theorem_score + assumptions_score + coverage_score + duplicate_score + parent_score
    rationale: list[str] = []

    if eligible_health["tailRatio"] < 0.6:
        score *= 0.7
        rationale.append("Eligible-set collapse detected in tail rounds.")

    tail8 = snapshots[-8:]
    if tail8 and all(float(item.get("eligibleCount", 0)) <= 1 for item in tail8):
        score *= 0.45
        rationale.append("Vacuous tail: only gateway remains eligible in final rounds.")

    if flapping_avg > 2.2:
        score *= 0.82
        rationale.append("High average flapping detected.")

    score = max(0.0, min(100.0, score))

    if score >= 80:
        verdict = "STABLE"
    elif score >= 60:
        verdict = "OSCILLATING"
    else:
        verdict = "UNSTABLE"

    return {
        "score": score,
        "verdict": verdict,
        "metrics": {
            "totalRounds": total_rounds,
            "theoremPassRate": theorem_pass_rate,
            "assumptionsPassRate": assumptions_pass_rate,
            "coverageAvg": coverage_avg,
            "duplicateEarly": duplicate_trend["early"],
            "duplicateLate": duplicate_trend["late"],
            "duplicateDrop": duplicate_trend["delta"],
            "parentChangeAvg": parent_change_avg,
            "flappingAvg": flapping_avg,
            "eligibleTailRatio": eligible_health["tailRatio"],
            "eligibleTailMin": eligible_health["minTail"],
            "eligibleTailMax": eligible_health["maxTail"],
        },
        "rationale": rationale,
    }
